### ISSUE TITLE
`background` builtin

### ISSUE THREAD

**Issue by nickthecook** [nickthecook/ops#22]
## Description

`ops` should support a `background` builtin that runs any action as a background task. E.g., with this action defined:

```yaml
actions:
  start:
    command: rackup
```

`ops start` will run `rackup` in the current terminal; the logs will be printed to the screen, and the user can CTRL+C to kill the app.

This is what the user would want if they were running the app directly, however:

- CI systems will need to start the app in the background sometimes; e.g. to run end-to-end tests
- [the `project dependency](#21) needs the ability to run an existing action as a background task

`ops background start` should run the action `start` as a background service. This may be accomplished using `screen`, `tmux`, `nohup`, etc.

Ideally, aliases for builtins could be implemented, so that `ops bg` functions like `ops background`.

## Interaction with user-defined actions

With this feature, the user can write `start`, `stop`, and `status`, and either run the app in the foreground (`ops start`) or the background (`ops background start`). If the `status` and `stop` actions are written properly, they should work whether the app is in the background or not, regardless of the underlying mechanism used by `ops` for backgrounding the app.

One action the user will not be able to write without an awareness of the underlying backgrounding mechanism is a `logs` action that is meant to dump the logs from the current or previous background session. `ops` may provide a builtin to accomplish this (e.g. `background-logs`; simply `logs` would be likely to conflict with user-defined actions), or it may choose to force the user to be aware of the backgrounding mechanism to do this themselves*.

**(The rest of this issue assumes `ops` will implement the `background-logs` feature)*

If aliases for builtins are implemented (as described above, so `ops bg` is an alias for `ops background`) then `ops bglogs` could be a more usable builtin than `background-logs`.

## Logs

The `background-logs` or `bglogs` builtin should take the name of an action, and dump the output of that action's background session to the terminal.

```yaml
actions:
  hello:
    command: echo HELLO && sleep 60 && echo GOODBYE
```

```shell
$ ops bg hello
Running `hello` in a background session...
$ ops bglogs hello
Displaying logs from background session for 'hello`...
HELLO
Background session 'hello' is still running.
$ # wait one minute
$ ops bglogs hello
HELLO
GOODBYE
Background session 'hello' exited with status 0.
$ ops bglogs nosuch
There is no background session for the action 'nosuch'.
$ 
```

## Follow logs

`ops bglogs -f hello` should follow the logs from the background session for `hello`.

```shell
$ ops bglogs -f hello
Following logs from background session for 'hello'...
HELLO
<terminal waits for more output until user presses CTRL+C or process terminates>
```

If the user enters any input, keystrokes are _not_ sent to the app. CTRL+C terminates the `ops bglogs -f` session, not the app.

## Attach

It should be possible for the user to attach their console to a background session.

```shell
$ ops attach hello
Attaching to background session for action 'hello'...
HELLO
<terminal waits for more output until user presses CTRL+C or process terminates>
```

If the user enters any input, it _is_ sent to the app. CTRL+C terminates the app. Any other signals generated from the keyboard are sent to the app.

**Comment by nickthecook** [comment]
0.9.0 includes a preliminary version of `ops background` and `ops bg`. It runs the given action or builtin in a subprocess and redirects output to a file:

<img width="1268" alt="image" src="https://user-images.githubusercontent.com/5835543/89924359-e979ac80-dbcf-11ea-8ecb-9fc53a0ad67b.png">

Differences between the issue description and this solution:

- ops is not aware of "background sessions", so no `attach` etc. builtins exist
- ops does not maintain separate logs per action; it stores only the last background log for the project

**Comment by nickthecook** [comment]
Still need to add:

- bglog
- documentation

**Comment by nickthecook** [comment]
Closed in 0.9.4.
